### PR TITLE
Complete Outline View - using dot parser lib and fallback to regex

### PR DESCRIPTION
**Pull request by tintinweb**
* This is a PoC and breaks some symbols functionality right now
* Using a 3rd party lib to parse the DOT file into an AST and building a vscode symbol tree from there

refs #95 

image

### CODE REVIEW

**Comment by tintinweb** [comment]
wdyt @bigbug?

added a new `DotParser.ts` that builds a vscode symbol tree for the outline. it currently breaks other symbol functionality. just created this as a PoC. would be great to get your input on whether we want to go forward using the parser for all symbol related things. thanks!

**Comment by bigbug** [comment]
I think the question here is what happens on syntax errors. Is the AST stable enough to handle this as well? The regex code is very fault tolerant.

**Comment by tintinweb** [comment]
In another extension I am only updating the AST on file change. If the parser bails we can show it in the problem view but keep the symbol tree. It will not be super accurate but should work most of the time. I agree that we should try to gratefully parse the code and not cause too much inconvenience while typing 🙌 let's explore this and decide if we want to use the parser.

**Comment by bigbug** [comment]
There seems to be a problem with html elements. See `example3.dot`

File: src/language/SymbolProvider.ts [hunk 1]
```diff
@@ -179,13 +181,21 @@ implements
...
     let symbols : SymbolInformation[] = [];
 
+    const dotParser = new DotParser();
+    dotParser.parse(document);
+    // console.log(parser.ast);
+    symbols = dotParser.getVscodeTypedAst().children;
+    // console.log(symbols);
+    /*
     symbols = symbols.concat(this.findExplicitNodeDefinition(document));
     symbols = symbols.concat(this.findNodeDefinition(document));
```
[the remark this answers is not in the document]
Comment (bigbug) [reply src/language/SymbolProvider.ts]: Fallback to regex reactivated for error throwing AST

**Comment by bigbug** [comment]
Current problem: The AST seems to not recognise nodes which have been defined within an edge:
Screenshot 2022-04-14 at 18 57 03

**Comment by bigbug** [comment]
DotParser now recognises edges and the nodes below.

**Comment by tintinweb** [comment]
looks good!

- we try to parse using dotparser, if that fails we fall back to the regex approach.
- the regex approach does not detect "digraph" elements yet. but that's fine
- we could push the dotparser syntax error to the diagnostics view although that might conflict with the web renderer that populates it right now. any thoughts?

going to merge this in the meantime 🥳

**Pull request merged**